streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java

```diff
@@ -703,16 +765,140 @@ public GroupAssignment assign(final Cluster metadata, final GroupSubscription gr
         return assignment;
     }
 
+    /**
+     * Generates an assignment that tries to satisfy two conditions: no active task previously owned by a consumer
+     * be assigned to another, and the number of tasks is evenly distributed throughout the client.
+     * <p>
+     * If it is impossible to satisfy both constraints we abort and return the interleaved assignment instead.
+     *
+     * @param metadata metadata for this client
+     * @param partitionsForTask mapping from task to its associated partitions
+     * @param allOwnedPartitions set of all partitions claimed as owned by the group
+     * @return task assignment for the consumers of this client
+     */
+    Map<String, List<TaskId>> giveTasksBackToConsumers(final ClientMetadata metadata,
```

Comment:
Added a new way to generate assignments that falls back to the old (interleaved) assignment if we can't avoid giving away a previously owned partition and triggering a second rebalance. I tried to comment it sufficiently well to explain what it's doing, but please give it a look and lmk if anything doesn't make sense